### PR TITLE
fix: check cached versions during transpile

### LINKED ISSUE #7759: cache invalidation does not work with --no-check
$DENO_DIR/gen cache is not being invalidated when a file changes and `--no-check` is specified:

```
/m/f/P/g/d/deno ❯❯❯ echo "console.log(42)" > abc.ts
/m/f/P/g/d/deno ❯❯❯ deno run --no-check abc.ts
42
/m/f/P/g/d/deno ❯❯❯ echo "console.log(43)" > abc.ts
/m/f/P/g/d/deno ❯❯❯ deno run --no-check abc.ts
42
```

cc @kitsonk - likely related to #7621

### PR DESCRIPTION
**Pull request by kitsonk**
Fixes #7759

This also includes a few other refactors that will make it easier to move other functions over to the new graph.

### CODE REVIEW

File: cli/tsc_config.rs [hunk 1]
```diff
@@ -202,12 +202,20 @@ pub fn parse_config(
 
 /// A structure for managing the configuration of TypeScript
 #[derive(Debug, Clone)]
-pub struct TsConfig(Value);
+pub struct TsConfig {
+  value: Value,
+  /// The bytes of any user config, which is used in generating hashes of
+  /// emitted files for cache invalidation purposes.
+  pub maybe_user_config: Option<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (kitsonk) [reply cli/tsc_config.rs:209]: But it isn't the `Value`, that is the final version of all the options, it is the bytes of a supplied `tsconfig` but actually you raise an interesting point, any change to occurs to the default options should invalidate as well. I will see how I can use the `Value`.

[the remark this answers is not in the document]
Comment (lucacasonato) [reply cli/tsc_config.rs:209]: Simplest is probably to hash the output of `serde_json::to_vec(value)`.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by ry** [comment]
no big deal but this commit appears to have caused an increase in max memory for the "hello" program
Screen Shot 2020-09-30 at 8 55 46 PM

**Comment by kitsonk** [comment]
> no big deal but this commit appears to have caused an increase in max memory for the "hello" program

Hmmmm... that is really strange, as it shouldn't have, as the new code should have only have impacted `no_check_hello` if any at all.  Certainly need to look into it.

**Comment by kitsonk** [comment]
@ry ok, I believe the reason is this:

- As @lucacasonato mentioned, the way we are generating the hash for the file is "improved" over the old way we do it (as it hashes the compilation settings), but this only occurs for `--no-check` at the moment.
- This means `deno run hello_002.ts` and `deno run --no-check hello_002.ts` will both think the cache is invalid when alternating between the two.
- We don't do a "dry run" of the `"hello"` benchmark, so we assumed it is already in the cache and _warm_, but `deno run hello_002.ts` sees it as an invalid cache and does a re-compile on the first benchmark run.
- The memory is a high water mark, while execution time is a mean and the first run is well outside the the standard deviation and so is discounted.

If we do a dry run of `"hello"` _before_ the benchmarking starts, I believe it will resolve the issue.  I will raise a PR to that effect.

**Comment by kitsonk** [comment]
Actually small augmentation to this:

* The hyperfine benchmarks run before the strace ones, the hyperfine "--no-check" occurs before the strace run, so those invalidate the hyperfine run causes the strace run to not be _warm_.

**Comment by kitsonk** [comment]
Fixed in #7770

Benchmarks___Deno

**Pull request merged**